### DISCUSSION TITLE
Build without runtime configuration (12 Factor)

### DISCUSSION THREAD

**Discussion by floriangosse** [vercel/next.js#46489] [Help]
I'm building an application that has a different build environment and runtime environment. And, I don't know the runtime configuration during build time as we follow the [12 Factor](https://12factor.net/) methodology.

Now we've introduced some validation and manipulation of configuration variables and it makes the build fail as the built files are executed by Next.js during the build time but the runtime configuration isn't available yet.

How are others solving this problem? Is there any workaround?

**Comment by icyJoseph** [comment]
I think someone on my team tried this once, with https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration, have you looked into it?

**Comment by floriangosse** [comment]
Yes, I'm using it quite heavily. But it doesn't solve the problem but just changes how I get the environment variables. Some of the configuration variables we're using are required to be available on the root scope of a file. But because Next.js is executing the files it fails as well.

**Comment by philwolstenholme** [comment]
Are these any help?

- https://github.com/vercel/next.js/discussions/17641, particularly https://github.com/vercel/next.js/discussions/17641#discussioncomment-2579143 and the comments underneath it
- https://github.com/sparanoid/env.sh
- https://github.com/andrewmclagan/react-env/

**Reply by floriangosse** [reply]
No. The problem I have isn't about how to get the environment variables into my application but that the environment variables aren't available during build time on purpose. The CI doesn't know the environment variables that are used during runtime as they have to change without a build.
